### DISCUSSION TITLE
Will Node 18 LTS become the default Node version on the agent images?

### DISCUSSION THREAD

**Discussion by TomasHubelbauer** [actions/virtual-environments#5429] [General questions] [answered]
Hi, Node 18 is out:
https://nodejs.org/en/blog/announcements/v18-release-announce

It will become LTS in October:
> Node.js 18 will be the 'Current' release for the next 6 months and then promoted to Long-term Support (LTS) in October 2022

At the same time, Node 16 will enter maintenance mode:
https://nodejs.org/en/about/releases (see chart)

It seems that GitHub Actions uses the current LTS as the Node version that's built-in to the agent images. I'm judging that based on announcements [like this one for Node 16](https://github.blog/changelog/2021-12-10-github-actions-github-hosted-runners-now-run-node-js-16-by-default).

Can we expect Node 18 will become the new default? And how do you estimate the delay will be between Node 18 reaching LTS status and GA shipping it as the default Node version? Thank you

**Comment by miketimofeev** [comment, marked as answer]
@TomasHubelbauer Hi!
Yes, once Node 18 becomes LTS we will switch the default version and announce it the same way as we did for Node 16
https://github.com/actions/virtual-environments/issues/4446

**Reply by TomasHubelbauer** [reply]
According to @MikeMcC399's  comment Node 18 is now cached on the Ubuntu latest agent image. Is there a date at which we can expect it will become the default Node version on the agent? According to https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md#language-and-runtime, Node 16.18.1 is still the default. Thanks!

**Reply by MikeMcC399** [reply]
@TomasHubelbauer 
If you write your own workflow and don't specify `uses:` [actions/setup-node@v3](https://github.com/actions/setup-node) to specify the version of Node.js to run under then you will get the default Node.js defined by the runner when running your workflow.

If you are talking about calling a pre-defined GitHub action in your workflow, for instance through `uses:` [actions/cache@v3](https://github.com/actions/cache) then it will use its own selection of Node.js in its context. See https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runs-for-javascript-actions with the example:
```yml
runs:
  using: 'node16'
  main: 'main.js'
```

I tripped up on this when I was looking at the [cypress-io/github-action](https://github.com/cypress-io/github-action/). See also issue https://github.com/cypress-io/github-action/issues/642 there.

In general it is probably a good idea to use the GitHub action `uses:` [actions/setup-node@v3](https://github.com/actions/setup-node) so that you do not rely on the default Node.js version.

**Reply by TomasHubelbauer** [reply]
I was specifically talking about running Node scripts in the workflow, not using Node-based actions. I am happy to use whatever the current default version of Node on the runner is, just curious when 18 will be made the default given it is the current LTS and I was under the impression that the latest agent image version will generally upgrade the installed Node version to the current LTS whenever that changes. I am aware I can specify my own preferred version of Node and that thanks to the caching this choice doesn't really add any runtime cost, but I prefer to stick with the defaults and keep my workflows simpler whenever possible.

**Reply by MikeMcC399** [reply]
@TomasHubelbauer 
Thanks for explaining! I understand where you are coming from now.

**Comment by MikeMcC399** [comment]
@miketimofeev 

Node 18 is now LTS - see https://nodejs.org/en/blog/release/v18.12.0/ from Oct 25, 2022. It would be great to see an announcement with your dates for switch-over.

**Reply by miketimofeev** [reply]
@al-cheb @mikhailkoliada @ddobranic ^^

**Reply by mikhailkoliada** [reply]
I will make all the preparations happen

**Reply by MikeMcC399** [reply]
I've noticed that Node.js 18 is now cached in the Ubuntu runners [20.04](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md#nodejs) and [22.04](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md#nodejs).

**Reply by MikeMcC399** [reply]
@mikhailkoliada 
You wrote back in October:
> I will make all the preparations happen

Did you get any further on these preparations?

**Reply by Vadorequest** [reply]
I also wish to upgrade to v18 directly, because it is quite time consuming to change all workflows of all my repositories, and I'd prefer to do it today for v18 rather than having to upgrade it again next year when v16 will be deprecated.

**Reply by MikeMcC399** [reply]
@Vadorequest 
According to https://github.com/nodejs/release#release-schedule both Node.js 14.x and 16.x will reach end-of-life this year (2023), so it would be good to make progress on Node.js 18.x soon. Perhaps @mikhailkoliada could give us some update on the plans? Node.js 20.x should also be coming in 3 months.

**Reply by MikeMcC399** [reply]
Is there any progress on the planning to make Node.js 18 the default version in the GitHub runners?

Looking back to the history, after Node.js 16 moved to (active) LTS status on Oct 26, 2021 it was made default in GitHub runners by Dec 2, 2021, according to https://github.com/actions/runner-images/issues/4446#issuecomment-984670008.

Node.js 18 moved to (active) LTS status on Oct 25, 2022 which is now almost 3 months ago.

Currently runners generally have Node.js 16.19.0 set as default (with the exception of [macOS-10.15](https://github.com/actions/runner-images/blob/main/images/macos/macos-10.15-Readme.md#language-and-runtime) which specifies 16.18.0).

**Reply by RaitoBezarius** [reply]
Hello there, I am a NixOS developer and we noticed that our packaging of GitHub Runner is using Node.js 16, we are EOLing Node.js 16 for the next release, as it doesn't seem possible to upgrade the GH Runner's Node.js without significant patching, we will mark it as insecure until this gets moved I believe.

**Reply by TomasHubelbauer** [reply]
@RaitoBezarius I was under the impression that Node 18 is now the default Node version? https://github.com/actions/runner-images/issues/7002 Curious how come you are seeing a different version. Or did I misunderstand?

**Comment by OR13** [comment]
wen

```yml
runs:
  using: 'node18'
  main: 'main.js'
```

`/me cries in crypto.subtle`

**Reply by MikeMcC399** [reply]
@OR13 

It sounds like you are introducing a separate topic regarding [`runs` for JavaScript actions](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runs-for-javascript-actions) which is currently available for  `node12` and `node16` only.

**Reply by OR13** [reply]
Yes, sorry, I misread the thread.

**Comment by MikeMcC399** [comment]
- https://github.com/actions/runner-images/issues/7002 now provides the date

**Comment by LucasFA** [comment]
For future reference and for any readers, finally it seems the transition is to node 20 by spring 2024 as per [this GitHub blog post](https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/)

**Comment by samus-aran** [comment]
We're locking this discussion because it has not had recent activity and/or other members have asked for more information to assist you but received no response. Thank you for helping us maintain a productive and tidy community for all our members.
